json.go

```diff
@@ -0,0 +1,809 @@
+		C.godror_dpiJsonArray_initialize((**C.dpiJsonNode)(unsafe.Pointer(&jsonnode)), C.uint32_t(len(arr)))
+
+		var dpijsonarr *C.dpiJsonArray
+		C.godror_dpiasJsonArray(jsonnode, (**C.dpiJsonArray)(unsafe.Pointer(&dpijsonarr)))
+		for index, entry := range arr {
+			var jsonnodelocal *C.dpiJsonNode
+			C.godror_setArrayElements(dpijsonarr, C.int(index), (**C.dpiJsonNode)(unsafe.Pointer(&jsonnodelocal)))
+			err := populateJsonNode(entry, jsonnodelocal)
+			if err != nil {
+				return err
+			}
+		}
+	case map[string]interface{}:
+		m, _ := in.(map[string]interface{})
+		// Initialize dpiJsonObjectNode
+		C.godror_dpiJsonObject_initialize((**C.dpiJsonNode)(unsafe.Pointer(&jsonnode)), C.uint32_t(len(m)))
+
+		var dpijsonobj *C.dpiJsonObject
+		C.godror_dpiasJsonObject(jsonnode, (**C.dpiJsonObject)(unsafe.Pointer(&dpijsonobj)))
+
+		var i C.int = 0
+		var cKey *C.char
+
+		for k, v := range m {
+			cKey = C.CString(k)
+			C.godror_dpiJsonObject_setKey(jsonnode, i, cKey, C.uint32_t(len(k)))
+			var jsonnodelocal *C.dpiJsonNode
+			C.free(unsafe.Pointer(cKey))
+			C.godror_setObjectFields(dpijsonobj, i, (**C.dpiJsonNode)(unsafe.Pointer(&jsonnodelocal)))
+			err := populateJsonNode(v, jsonnodelocal)
+			if err != nil {
+				return err
+			}
+			i = i + 1
+		}
+
+	case int:
+		C.godror_dpiJson_setInt64(jsonnode, C.int64_t(x))
+	case int8:
+		C.godror_dpiJson_setInt64(jsonnode, C.int64_t(x))
+	case int16:
+		C.godror_dpiJson_setInt64(jsonnode, C.int64_t(x))
+	case int32:
+		C.godror_dpiJson_setInt64(jsonnode, C.int64_t(x))
+	case int64:
+		C.godror_dpiJson_setInt64(jsonnode, C.int64_t(x))
+	case uint:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case uint8:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case uint16:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case uint32:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case uint64:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case float32:
+		C.godror_dpiJson_setDouble(jsonnode, C.double(x))
+	case float64:
+		C.godror_dpiJson_setDouble(jsonnode, C.double(x))
+	case string:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setString(jsonnode, &(data.dpiData))
+	case time.Time:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setTime(jsonnode, &(data.dpiData))
+	case time.Duration:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setIntervalDS(jsonnode, &(data.dpiData))
+	case []byte:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setBytes(jsonnode, &(data.dpiData))
+	case bool:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setBool(jsonnode, &(data.dpiData))
+	default:
+		return fmt.Errorf("Unsupported type %T\n", in)
+	}
+	return nil
+}
+
+// Creates a JSONInt from user input []interface{}
+func newJSONScalar(val interface{}, jscalar *JSONScalar) error {
```

Comment:
Why the jscalar pointer? Why not just return *JSONScalar?

Reply:
I didnt get it completely. You mean something like below, I submitted change in this way. Please suggest:
return &JSONScalar{..

Reply:
Fixed.